stac_fastapi/pgstac/stac_fastapi/pgstac/db.py

```diff
@@ -0,0 +1,141 @@
+"""Database connection handling."""
+import logging
+
+import orjson
+from asyncpg import exceptions
+from buildpg import asyncpg, render
+from fastapi import FastAPI
+
+from stac_fastapi.types.errors import (
+    ConflictError,
+    DatabaseError,
+    ForeignKeyError,
+    NotFoundError,
+)
+
+# from stac_fastapi.pgstac.config import Settings
+# settings = Settings()
+# if app.state.testing:
+#     settings.reader_connection_string = settings.writer_connection_string = settings.testing_connection_string
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+async def con_init(conn):
+    """Use orjson for json returns."""
+    await conn.set_type_codec(
+        "json",
+        encoder=orjson.dumps,
+        decoder=orjson.loads,
+        schema="pg_catalog",
+    )
+    await conn.set_type_codec(
+        "jsonb",
+        encoder=orjson.dumps,
+        decoder=orjson.loads,
+        schema="pg_catalog",
+    )
+
+
+async def connect_to_db(app: FastAPI) -> None:
+    """Connect to Database."""
+    settings = app.state.settings
+    if app.state.settings.testing:
+        readpool = writepool = settings.testing_connection_string
+    else:
+        readpool = settings.reader_connection_string
+        writepool = settings.writer_connection_string
+    logger.info(f"Creating Connection Pools {readpool} {writepool}")
+    db = DB()
+    app.state.readpool = await db.create_pool(readpool, settings)
+    app.state.writepool = await db.create_pool(writepool, settings)
+
+
+async def close_db_connection(app: FastAPI) -> None:
+    """Close connection."""
+    logger.info("Closing connections to database")
+    await app.state.readpool.close()
+    await app.state.writepool.close()
+    logger.info("Connections closed")
+
+
+async def dbfunc(pool, func, arg):
+    """Wrap PLPGSQL Functions."""
+    try:
+        if isinstance(arg, str):
+            async with pool.acquire() as conn:
+                q, p = render(
+                    f"""
+                        SELECT * FROM {func}(:item::text);
+                        """,
+                    item=arg,
+                )
+                return await conn.fetchval(q, *p)
+        else:
+            async with pool.acquire() as conn:
+                q, p = render(
+                    f"""
+                        SELECT * FROM {func}(:item::text::jsonb);
+                        """,
+                    item=arg.json(),
+                )
+                return await conn.fetchval(q, *p)
+    except exceptions.UniqueViolationError:
```

Comment:
We should pass through the traceback from the initial exception so it doesn't get lost, this should make it easier to debug database related errors.

```python
except UniqueViolationError as e:
    raise ConflictError from e
```